File: web_server.py
import http.server
import socketserver
import os
import sys
import webbrowser
import threading
import time

class MyHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_header('Cross-Origin-Opener-Policy', 'same-origin')
        self.send_header('Cross-Origin-Embedder-Policy', 'require-corp')
        super().end_headers()

def open_browser(port):
    time.sleep(1.2)
    url = f'http://localhost:{port}'
    print(f'\n[✓] 正在自动打开浏览器: {url}')
    webbrowser.open(url)

def main():
    # 切换工作目录到 build/web
    base_dir = os.path.dirname(os.path.abspath(__file__))
    web_dir = os.path.join(base_dir, 'mobile_agent', 'build', 'web')
    
    if not os.path.exists(web_dir):
        # 兼容在 mobile_agent 目录下运行
        web_dir = os.path.join(base_dir, 'build', 'web')
    
    if not os.path.exists(web_dir):
        print(f'[!] 错误: 未找到 Web 产物目录 ({web_dir})')
        print('[!] 请先运行 flutter build web 编译项目！')
        input('按回车键退出...')
        sys.exit(1)
        
    os.chdir(web_dir)
    port = 8080
    
    for p in range(8080, 8090):
        try:
            handler = MyHTTPRequestHandler
            socketserver.TCPServer.allow_reuse_address = True
            httpd = socketserver.TCPServer(('', p), handler)
            port = p
            break
        except OSError:
            continue

    print('=' * 55)
    print('       🚀 NEXUS Agent Web 本地服务已启动')
    print('=' * 55)
    print(f' * 访问地址: http://localhost:{port}')
    print(' * Web WASM 跨域隔离支持: 已启用 (COOP/COEP)')
    print(' * 提示: 按 Ctrl + C 可停止服务')
    print('=' * 55)

    threading.Thread(target=open_browser, args=(port,), daemon=True).start()

    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print('\n服务已停止。')
        httpd.server_close()

if __name__ == '__main__':
    main()

File: web_server.dart
import 'dart:io';
import 'dart:async';

void main() async {
  Directory dir = Directory('mobile_agent/build/web');
  if (!dir.existsSync()) {
    dir = Directory('build/web');
  }
  if (!dir.existsSync()) {
    stdout.writeln('[!] 错误: 未找到 Web 产物目录，请先编译项目！');
    exit(1);
  }

  HttpServer? server;
  int port = 8080;
  for (var p = 8080; p < 8090; p++) {
    try {
      server = await HttpServer.bind(InternetAddress.loopbackIPv4, p);
      port = p;
      break;
    } catch (_) {}
  }

  if (server == null) {
    stdout.writeln('[!] 无法绑定端口 8080-8090');
    exit(1);
  }

  stdout.writeln('=======================================================');
  stdout.writeln('       🚀 NEXUS Agent Web 本地服务已启动');
  stdout.writeln('=======================================================');
  stdout.writeln(' * 访问地址: http://localhost:$port');
  stdout.writeln(' * Web WASM 跨域隔离支持: 已启用 (COOP/COEP)');
  stdout.writeln(' * 提示: 按 Ctrl + C 可停止服务');
  stdout.writeln('=======================================================');

  Timer(const Duration(milliseconds: 1200), () {
    final url = 'http://localhost:$port';
    stdout.writeln('\n[✓] 正在自动打开浏览器: $url');
    if (Platform.isWindows) {
      Process.run('cmd', ['/c', 'start', '', url]);
    } else if (Platform.isMacOS) {
      Process.run('open', [url]);
    } else if (Platform.isLinux) {
      Process.run('xdg-open', [url]);
    }
  });

  await for (HttpRequest request in server) {
    request.response.headers.set('Cross-Origin-Opener-Policy', 'same-origin');
    request.response.headers
        .set('Cross-Origin-Embedder-Policy', 'require-corp');
    request.response.headers
        .set('Cache-Control', 'no-cache, no-store, must-revalidate');

    String reqPath = request.uri.path;
    if (reqPath == '/' || reqPath.isEmpty) reqPath = '/index.html';

    final cleanPath = reqPath.split('?').first;
    var file = File('${dir.path}$cleanPath');

    if (!await file.exists()) {
      if (!cleanPath.contains('.')) {
        file = File('${dir.path}/index.html');
      }
    }

    if (await file.exists()) {
      final ext = file.path.split('.').last.toLowerCase();
      switch (ext) {
        case 'html':
          request.response.headers.contentType = ContentType.html;
          break;
        case 'js':
        case 'mjs':
          request.response.headers.contentType =
              ContentType('application', 'javascript', charset: 'utf-8');
          break;
        case 'wasm':
          request.response.headers.contentType =
              ContentType('application', 'wasm');
          break;
        case 'css':
          request.response.headers.contentType =
              ContentType('text', 'css', charset: 'utf-8');
          break;
        case 'json':
          request.response.headers.contentType = ContentType.json;
          break;
        case 'png':
          request.response.headers.contentType = ContentType('image', 'png');
          break;
        case 'jpg':
        case 'jpeg':
          request.response.headers.contentType = ContentType('image', 'jpeg');
          break;
        case 'svg':
          request.response.headers.contentType =
              ContentType('image', 'svg+xml');
          break;
        case 'ico':
          request.response.headers.contentType = ContentType('image', 'x-icon');
          break;
        case 'otf':
          request.response.headers.contentType = ContentType('font', 'otf');
          break;
        case 'ttf':
          request.response.headers.contentType = ContentType('font', 'ttf');
          break;
      }
      await file.openRead().pipe(request.response);
    } else {
      request.response.statusCode = HttpStatus.notFound;
      request.response.write('404 Not Found');
      await request.response.close();
    }
  }
}
